src/storage/src/compacter.rs

```diff
@@ -0,0 +1,691 @@
+                // We might get a lot of messages for relations we don't
+                // know about here so ignore those.
+                if let Some(trace) = self.traces.get_mut(&id) {
+                    if let Some(compaction_frontier) = trace.compaction {
+                        assert!(frontier >= compaction_frontier);
+                    }
+                    trace.compaction = Some(frontier);
+                }
+            }
+        };
+        Ok(())
+    }
+
+    pub async fn run(&mut self) {
+        let ret = self.compact().await;
+
+        match ret {
+            Ok(_) => (),
+            Err(e) => {
+                error!("Compacter thread encountered an error: {:#}", e);
+                error!("Shutting down compacter thread. No further updates will be persisted.");
+            }
+        }
+    }
+}
+
+/// Read a directory and return all (non-subdirectory) files matching `regex`.
+fn read_dir_regex(path: &Path, regex: &Regex) -> Result<Vec<PathBuf>, anyhow::Error> {
+    let entries = std::fs::read_dir(path).with_context(|| {
+        format!(
+            "failed to read {} looking for {}",
+            path.display(),
+            regex.as_str()
+        )
+    })?;
+    let mut results = vec![];
+    for entry in entries {
+        if let Ok(file) = entry {
+            let path = file.path();
+            let file_name = path.file_name();
+            if file_name.is_none() {
+                continue;
+            }
+
+            let file_name = file_name.unwrap().to_str();
+
+            if file_name.is_none() {
+                continue;
+            }
+
+            let file_name = file_name.unwrap();
+            if regex.is_match(&file_name) {
+                results.push(path.to_path_buf());
+            }
+        }
+    }
+
+    Ok(results)
+}
+
+// Data stored in Batches
+#[derive(Debug, PartialEq)]
+pub enum Message {
+    // (Row, time, diff) tuples
+    Data(Update),
+    // Statements about which timestamps we might still receive data at.
+    Progress(Timestamp),
+}
+
+fn read_message(buf: &[u8], mut offset: usize) -> Option<(Message, usize)> {
+    if offset >= buf.len() {
+        return None;
+    }
+
+    // Let's start by only looking at the buffer at the offset.
+    let (_, data) = buf.split_at(offset);
+
+    if data.len() < 12 {
+        error!(
+            "invalid offset while reading file: {}. Expected at least 12 more bytes have {}",
+            offset,
+            data.len()
+        );
+        return None;
+    }
+
+    // Let's read the header first
+    let mut cursor = Cursor::new(&data);
+
+    let is_progress = cursor.read_u32::<NetworkEndian>().unwrap();
+
+    if is_progress != 0 {
+        // If this is a progress message let's seal a new
+        // set of updates.
+
+        // Lets figure out the time bound.
+        let timestamp = cursor.read_u64::<NetworkEndian>().unwrap();
+        // Advance the offset past what we've read.
+        offset += 12;
```

Comment:
Is the 12 equal to 4 from before and now 8 again? It could perhaps be done closer to the cursor change otherwise I foresee refactoring pain.

Another way to structure this, fwiw, is to do all of your business and then check `Cursor::position()` at the end to see what you consumed. Or, maybe it is hard for some other reason, but the manual tracking of offsets not always adjacent to where they are changed seems scary.